### PR TITLE
Fixes #1361. Private fields promotion tests added

### LINKED ISSUE #1361: Add tests for private fields promotion
See https://github.com/dart-lang/language/issues/2020 for specification

### PR DESCRIPTION
**Pull request by sgrekhov**

### CODE REVIEW

File: LanguageFeatures/Private-fields-promotion/promotion_A01_t05.dart [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+  A(this._x);
+}
+
+class C extends A {
+  final int? _x;
+  C(this._x) : super(_x);
+
+  void testC() {
+    if (_x != null) {
+      super._x.isOdd;
```
Comment (stereotype441) [inline LanguageFeatures/Private-fields-promotion/promotion_A01_t05.dart:28]: Ooh, nice corner case!  My current prototype code doesn't handle it properly.  I'll make sure to fix it :)

File: LanguageFeatures/Private-fields-promotion/promotion_A05_t04.dart [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+/// @assertion An instance field is promotable only if
+/// (a) it is private,
+/// (b) it is final, and
+/// (c) all other concrete instance getters with the same name in the same
+/// library are also final fields
+///
+/// @description Checks that if there are other concrete instance getters with
+/// the same name in the same library and they are not a final fields then the
+/// field is not promotable. Test the case when there is a getter with the same
+/// name in some extension in the same library
```
Comment (stereotype441) [inline LanguageFeatures/Private-fields-promotion/promotion_A05_t04.dart:14]: Hmm, I had not thought about this case.  I'm ok with leaving the test as is for now, but I've filed https://github.com/dart-lang/language/issues/2404 to check if anyone on the language team feels strongly that this should behave differently.

**Review by stereotype441 — APPROVED** [review]
LGTM, thank you!

**Comment by sgrekhov** [comment]
@eernstg please review

File: LanguageFeatures/Private-fields-promotion/promotion_A01_t05.dart [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+//       ^^^^^
+// [analyzer] unspecified
+// [cfe] unspecified
+    }
+  }
+}
+
+main() {
+  C(42).testC();
+}
```
Comment (eernstg) [inline LanguageFeatures/Private-fields-promotion/promotion_A01_t05.dart:63]: Cool!

File: LanguageFeatures/Private-fields-promotion/promotion_A05_t05.dart [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+      _x.isOdd;
+//       ^^^^^
+// [analyzer] unspecified
+// [cfe] unspecified
+    }
+  }
+}
+
+class C {
+  int? get _x => 43;
```
Comment (eernstg) [inline LanguageFeatures/Private-fields-promotion/promotion_A05_t05.dart:33]: For all these conflicts, perhaps test that there is no conflict when the "other" declaration that creates the conflict is `static`?

File: LanguageFeatures/Private-fields-promotion/promotion_A05_t08.dart [hunk 5]
```diff
@@ -0,0 +1,40 @@
...
+
+// SharedOptions=--enable-experiment=inference-update-2
+
+class A<T> {
+  T _x;
+  A(this._x);
+
+  void testA() {
+    if (_x is int) {
+      _x.isOdd;
```
Comment (eernstg) [inline LanguageFeatures/Private-fields-promotion/promotion_A05_t08.dart:25]: This is probably just testing that we get a non-promotable getter when `A._x` isn't final (so it's probably an essentially-duplicate of an earlier test, and class `C` makes no difference ... but it isn't wrong).

File: LanguageFeatures/Private-fields-promotion/promotion_A05_t10.dart [hunk 6]
```diff
@@ -0,0 +1,39 @@
...
+/// the same name in the same library and they are not a final fields then the
+/// field is not promotable. Test when the variable with the same name is not
+/// final in some class in the same library but in another file
+/// @author sgrekhov22@gmail.com
+
+// SharedOptions=--enable-experiment=inference-update-2
+
+library promotion_A05_t10;
+
+part "promotion_A05_t10.lib.dart";
```
Comment (eernstg) [inline LanguageFeatures/Private-fields-promotion/promotion_A05_t10.dart:21]: Good catch, thanks!

File: LanguageFeatures/Private-fields-promotion/promotion_A05_t05.dart (hunk 4 above)
Comment (eernstg) [reply LanguageFeatures/Private-fields-promotion/promotion_A05_t05.dart:33]: Resolving: I think this is covered below already.

**Review by eernstg — APPROVED** [review]
Looks good!

**Pull request merged**